common/src/main/java/io/netty/util/NetUtil.java

```diff
@@ -316,101 +316,113 @@ public static boolean isIpV6AddressesPreferred() {
+        // insert 0s.
+
+        // Go through the tokens, including the separators ':' and '.'
+        // When we hit a : or . the previous token will be added to either
+        // the hex list or decimal list. In the case where we hit a ::
+        // we will save the index of the hexStrings so we can add zeros
+        // in to fill out the string
+        while (tokenizer.hasMoreTokens()) {
+            prevToken = token;
+            token = tokenizer.nextToken();
+
+            if (":".equals(token)) {
+                if (":".equals(prevToken)) {
                     doubleColonIndex = hexStrings.size();
-                } else {
-                    hexStrings.add(token);
+                } else if (!prevToken.isEmpty()) {
+                    hexStrings.add(prevToken);
                 }
-            } else if (".".equals(prevToken)) {
-                decStrings.add(token);
+            } else if (".".equals(token)) {
+                decStrings.add(prevToken);
             }
+        }
 
-            // figure out how many hexStrings we should have
-            // also check if it is a IPv4 address
-            int hexStringsLength = 8;
-
-            // If we have an IPv4 address tagged on at the end, subtract
-            // 4 bytes, or 2 hex words from the total
-            if (!decStrings.isEmpty()) {
-                hexStringsLength -= 2;
+        if (":".equals(prevToken)) {
+            if (":".equals(token)) {
+                doubleColonIndex = hexStrings.size();
+            } else {
+                hexStrings.add(token);
             }
+        } else if (".".equals(prevToken)) {
+            decStrings.add(token);
+        }
 
-            // if we hit a double Colon add the appropriate hex strings
-            if (doubleColonIndex != -1) {
-                int numberToInsert = hexStringsLength - hexStrings.size();
-                for (int i = 0; i < numberToInsert; i ++) {
-                    hexStrings.add(doubleColonIndex, "0");
-                }
-            }
+        // figure out how many hexStrings we should have
+        // also check if it is a IPv4 address
+        int hexStringsLength = 8;
 
-            byte[] ipByteArray = new byte[IPV6_BYTE_COUNT];
+        // If we have an IPv4 address tagged on at the end, subtract
+        // 4 bytes, or 2 hex words from the total
+        if (!decStrings.isEmpty()) {
+            hexStringsLength -= 2;
+        }
 
-            // Finally convert these strings to bytes...
-            for (int i = 0; i < hexStrings.size(); i ++) {
-                convertToBytes(hexStrings.get(i), ipByteArray, i << 1);
+        // if we hit a double Colon add the appropriate hex strings
+        if (doubleColonIndex != -1) {
+            int numberToInsert = hexStringsLength - hexStrings.size();
+            for (int i = 0; i < numberToInsert; i ++) {
+                hexStrings.add(doubleColonIndex, "0");
```

Comment:
did you consider just accounting for this directly in `ipByteArray` instead of adding to `hexStrings` first? You could save some operations on the collection.